anchor/programs/anchor-agnostic-orderbook/src/lib.rs

```diff
@@ -0,0 +1,492 @@
+            &ctx.accounts.bids,
+            &ctx.accounts.asks,
+            market_state.callback_info_len as usize,
+            market_state.callback_id_len as usize,
+        )
+        .unwrap();
+        if !orderbook_state.is_empty() {
+            msg!("The orderbook must be empty");
+            return Err(ProgramError::from(AoError::MarketStillActive));
+        }
+
+        // Check if all events have been processed
+        let header = {
+            let mut event_queue_data: &[u8] =
+                &ctx.accounts.event_queue.data.borrow()[0..EVENT_QUEUE_HEADER_LEN];
+            EventQueueHeader::deserialize(&mut event_queue_data).unwrap()
+        };
+        if header.count != 0 {
+            msg!("The event queue needs to be empty");
+            return Err(ProgramError::from(AoError::MarketStillActive));
+        }
+
+        market_state.tag = AccountTag::Uninitialized as u64;
+        let market = ctx.accounts.market.to_account_info();
+        let event_queue = ctx.accounts.event_queue.to_account_info();
+        let bids = ctx.accounts.bids.to_account_info();
+        let asks = ctx.accounts.asks.to_account_info();
+        let lamports_target_account = ctx.accounts.lamports_target_account.to_account_info();
+
+        let mut market_lamports = market.try_borrow_mut_lamports()?;
+        let mut event_queue_lamports = event_queue.try_borrow_mut_lamports()?;
+        let mut bids_lamports = bids.try_borrow_mut_lamports()?;
+        let mut asks_lamports = asks.try_borrow_mut_lamports()?;
+        let mut target_lamports = lamports_target_account.try_borrow_mut_lamports()?;
+
+        **target_lamports +=
+            **market_lamports + **bids_lamports + **asks_lamports + **event_queue_lamports;
+
+        **market_lamports = 0;
+        **bids_lamports = 0;
+        **asks_lamports = 0;
+        **event_queue_lamports = 0;
+
+        orderbook_state.release(&ctx.accounts.bids, &ctx.accounts.asks);
+
+        Ok(())
+    }
+}
+
+#[derive(Accounts)]
+pub struct CreateMarket<'info> {
+    // TODO PDAs?
+    #[account(init, payer = payer)]
+    pub market: Account<'info, MarketState>,
+    // TODO pass in space size instead of just max
+    #[account(init, payer = payer, space = 10240)]
+    pub event_queue: AccountInfo<'info>,
+    // TODO it would be nicer to parameterize with the actual types `T` instead of the `AccountInfo`
+    // escape hatch.
+    //
+    // We want to let Anchor do the serde boilerplate, checks etc. for us
+    // but `Slab` does not implement `Clone` (which Anchor needs -- see comment on `Slab` struct)
+    #[account(init, payer = payer, space = 10240)]
+    pub bids: AccountInfo<'info>,
+    #[account(init, payer = payer, space = 10240)]
+    pub asks: AccountInfo<'info>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct NewOrder<'info> {
+    #[account(mut)]
+    pub market: Account<'info, MarketState>,
```

Comment:
I think it's definitely possible to make the `MarketState` zero-copy by using `AccountLoader`